src/wasm/wasm-type.cpp

```diff
@@ -14,38 +14,193 @@
  * limitations under the License.
  */
 
+#include <cassert>
+#include <shared_mutex>
+#include <sstream>
+#include <unordered_map>
+
 #include "wasm-type.h"
 #include "wasm-features.h"
 
 #include "compiler-support.h"
-#include <cstdlib>
+
+template<> class std::hash<std::vector<wasm::Type>> {
+public:
+  size_t operator()(const std::vector<wasm::Type>& types) const {
+    size_t res = 0;
+    for (auto vt : types) {
+      res ^= hash<uint32_t>{}(vt);
```

Comment:
where is this `hash` from? (we have `src/support/hash.h` which ensures the same hash values across libcs, but this doesn't look like it?)

Reply:
This is `std::hash`. I will add the namespace to disambiguate for the reader.